### PR TITLE
[homematic] Validate datapoint values before writing to config

### PR DESCRIPTION
**Pull request by maniac103**
HM config datapoint values on some devices can be out of their valid
range in some cases, particularly if they are unused by the device
currently. Since openhab-core now validates attempts to update
configuration by the thing handlers, make sure we always send a valid
configuration even for those affected datapoints.

One example for such behaviour is HmIPW-DRBL4, which has multiple datapoints which depend on each other:
- POWERUP_JUMPTARGET can have values OFF, ON, OFF_DELAY, ON_DELAY
- POWERUP_ONDELAY_VALUE, POWERUP_ONDELAY_UNIT are only used if POWERUP_JUMPTARGET has value ON_DELAY
- likewise, POWERUP_OFFDELAY_VALUE and POWERUP_OFFDELAY_UNIT are only used if POWERUP_JUMPTARGET has value OFF_DELAY
- if e.g. the POWERUP_JUMPTARGET is OFF, e.g. POWERUP_ONDELAY_VALUE might stay uninitialized by CCU and/or device itself, in which case it may be reported with an invalid value

Fixes the issue reported [here](https://github.com/openhab/openhab-addons/pull/12192#issuecomment-1080436944)

### CODE REVIEW

**Comment by maniac103** [comment]
@fwolter @MHerbst Please review as well ;-)

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java [hunk 1]
```diff
@@ -420,6 +419,37 @@ protected void updateDatapointState(HmDatapoint dp) {
...
+        if (dp.isNumberType()) {
+            // For number datapoints that are only used depending on the value of other datapoints,
+            // the CCU may return invalid (out of range) values if the datapoint currently is not in use.
+            // Make sure to not invalidate the whole configuration by returning the datapoint's default
+            // value in that case.
+            final boolean minValid, maxValid;
+            if (dp.isFloatType()) {
+                Double numValue = dp.getDoubleValue();
+                minValid = dp.getMinValue() == null || numValue >= dp.getMaxValue().doubleValue();
+                maxValid = dp.getMinValue() == null || numValue <= dp.getMinValue().doubleValue();
```
Comment (lolodomo) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java:438]: It looks to me wrong: the first should I believe consider dp.getMinValue() while the second should consider dp.getMaxValue(). WDYT ?
Same remark in the else branch

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java [hunk 2]
```diff
@@ -140,8 +140,7 @@ private void doInitializeInBackground() throws GatewayNotAvailableException, Hom
             loadHomematicChannelValues(channel);
             for (HmDatapoint dp : channel.getDatapoints()) {
                 if (dp.getParamsetType() == HmParamsetType.MASTER) {
-                    config.put(MetadataUtils.getParameterName(dp),
-                            dp.isEnumType() ? dp.getOptionValue() : dp.getValue());
+                    config.put(MetadataUtils.getParameterName(dp), getValueForConfiguration(dp));
```
Comment (lolodomo) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java:143]: As getValueForConfiguration can retrun null, I am not sure whether this config entry should be set to null, or not set at all ?
I am even not sure if this is valid to set a null value for a config parameter.

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDatapoint.java [hunk 3]
```diff
@@ -150,19 +150,34 @@ public int getOptionIndex(String option) {
      */
     public String getOptionValue() {
         if (options != null && value != null) {
-            int idx = 0;
-            if (value instanceof Integer) {
-                idx = (int) value;
-            } else {
-                idx = Integer.parseInt(value.toString());
-            }
+            int idx = getIntegerValue();
```
Comment (lolodomo) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDatapoint.java:153]: As getIntegerValue can return null, there is probably a test to include here before casting it to an int.

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java [hunk 4]
```diff
@@ -420,6 +419,37 @@ protected void updateDatapointState(HmDatapoint dp) {
         }
     }
 
+    private Object getValueForConfiguration(HmDatapoint dp) {
```
Comment (lolodomo) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java:422]: Please add @Nullable to the return type.

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDatapoint.java [hunk 5]
```diff
@@ -150,19 +150,34 @@ public int getOptionIndex(String option) {
...
             if (idx < options.length) {
                 return options[idx];
             }
         }
         return null;
     }
 
+    public Integer getIntegerValue() {
```
Comment (lolodomo) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDatapoint.java:161]: Please add @Nullable to the return type.

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDatapoint.java [hunk 6]
```diff
@@ -150,19 +150,34 @@ public int getOptionIndex(String option) {
...
+        if (value instanceof Integer) {
+            return (int) value;
+        } else if (value != null) {
+            return Integer.parseInt(value.toString());
+        } else {
+            return null;
+        }
+    }
+
+    public Double getDoubleValue() {
```
Comment (lolodomo) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDatapoint.java:171]: Please add @Nullable to the return type.

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java (hunk 1 above)
Comment (maniac103) [reply bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java]: Yes, you're completely right. Not sure what I was thinking there, probably copy'n'paste gone wrong. Fixed now.

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDatapoint.java (hunk 3 above)
Comment (maniac103) [reply bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDatapoint.java]: Added null check

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java (hunk 4 above)
Comment (maniac103) [reply bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java]: I've done so (same as in the occasions below), but I guess this binding will need a lot more work before being fully nullability-correct.

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java (hunk 2 above)
Comment (maniac103) [reply bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java:144]: The only case under which `getValueForConfiguration()` returns null is if the datapoint holds a `null` value. In that case I'd say that writing `null` into the config is correct.
The places I checked that care about the keys present in the map (those that call `containsKey()` and `keySet()` seem to be able to deal with null values being present. Not sure about e.g. the UI, but given the previous code put `null` values in this case (value being `null`) as well, I guess it should be fine.

**Review by lolodomo — APPROVED** [review]
LGTM

**Pull request merged**